File: 백준/Silver/15649. N과 M （1）/N과 M （1）.py
# n개 중에서 m개를 뽑는 조합이므로 백트래킹을 사용한다.
# 백트래킹은 모든 가능한 경우의 수 중에 특정한 조건을 만족하는 경우만 살피는 것이다.
# 재귀 함수를 이용한 백트래킹 함수를 만들어 사용한다.
# answer에 str(i)로 append했더니 이미 뽑힌 수임에도 아니라고 판명나서 출력오류..

import sys

n, m = map(int, sys.stdin.readline().split())
answer = []

def backTracking() :
    if len(answer) == m : print(" ".join(map(str, answer))); return # m개만큼 뽑으면 그만

    for i in range(1, n+1) :
        if i not in answer : # 이미 뽑힌 수가 아니라면
            answer.append(i) # 답에 집어넣는다
            backTracking() # 재귀함수 호출
            answer.pop() # 자식 노드를 돌고 지움 > 자식 노드를 다 돌고나면 부모 노드 지움
backTracking()

File: 백준/Silver/15649. N과 M （1）/N과 M （1）.java
/*
 * 백트래킹-재귀를 이용해서 푼다.
 * 출력하기 전, 수열을 보관할 큐를 생성한다 => 배열은 길이가 정해져 있기 때문
 * 1부터 n+1까지 반복문을 돌며, i가 큐에 없다면 집어넣고 재귀, 있다면 continue
 * 큐의 크기가 m과 같다면 출력하고, 큐에서 해당 값을 제거한다.
 * 반복문 안에서 재귀를 부르기 때문에, 전부 다 불린 숫자는 해당 자리에서 다시 나올 수 없음
 * 큐라고 하지만 Dequeue 사용 => 스택처럼 후입선출
 */

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.ArrayDeque;
import java.util.Deque;

public class Main {
	public static int N; // 1부터 N까지의 숫자 => 범위
	public static int M; // 만족해야 하는 수열의 개수
	public static Deque<Integer> sequence; // 1~N개의 숫자 중 M개를 보관할 Dequeue
	public static StringBuilder sb; // 함수에서 쓰기 위한 StringBuilder
	
	public static void main(String[] args) throws NumberFormatException, IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in)); // 입력
		BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out)); // 출력
		String []input = br.readLine().split(" "); // 입력값 배열
		N = Integer.parseInt(input[0]); // 정수로 변환
		M = Integer.parseInt(input[1]); // 정수로 변환
		sequence = new ArrayDeque<Integer>(); // 배열 초기화
		sb = new StringBuilder(40320); // 최댓값이 8이므로 8!까지 수용 가능한 Builder
		
		BackTracking(); // 함수 호출
		bw.append(sb); // bw에 sb 추가
		bw.flush(); // bw에 남은 글자 모두 출력
		br.close(); // 입력 종료
		bw.close(); // 출력 종료
	}
	
	public static void BackTracking() throws IOException {
		if(sequence.size() == M) { // 큐의 크기가 M과 같다면
			for(int q : sequence){ // 0부터 M까지
				sb.append(q); // bw에 수열 추가
				sb.append(" ");
			}
			sb.append("\n");
			return; // 재귀 종료
		}
		for(int i=1; i<N+1; i++) {
			if (!sequence.contains(i)) { // 만약 큐에 i가 없다면
				sequence.addLast(i); // 큐 뒤에 추가
				BackTracking(); // 다시 재귀 => 다음 자리수 찾으러 Go!
				sequence.removeLast(); // 해당 자리수에 i가 있는 경우의 모두 탐색 => i 제거
			}
		}
	}
}
